sql/src/test/java/io/crate/expression/scalar/systeminformation/PgTypeofFunctionTest.java

```diff
@@ -0,0 +1,209 @@
+
+import io.crate.expression.scalar.AbstractScalarFunctionsTest;
+import io.crate.expression.symbol.Literal;
+import io.crate.types.DataTypes;
+import io.crate.types.ObjectType;
+import org.junit.Test;
+
+import java.util.List;
+
+public class PgTypeofFunctionTest extends AbstractScalarFunctionsTest {
+
+    @Test
+    public void test_primitive_types() {
+        assertEvaluate("pg_typeof(null)", DataTypes.UNDEFINED.getName());
+        assertEvaluate("pg_typeof(null::bigint)", DataTypes.LONG.getName());
+        assertEvaluate("pg_typeof(name)", DataTypes.STRING.getName(), Literal.of((String) null));
+
+        assertEvaluate("pg_typeof(true)", DataTypes.BOOLEAN.getName());
+        assertEvaluate("pg_typeof(is_awesome)", DataTypes.BOOLEAN.getName(), Literal.of(true));
+
+        assertEvaluate("pg_typeof(58::char)", DataTypes.BYTE.getName());
+        assertEvaluate("pg_typeof(c)", DataTypes.BYTE.getName(), Literal.of(58));
+
+        assertEvaluate("pg_typeof(10::smallint)", DataTypes.SHORT.getName());
+        assertEvaluate("pg_typeof(short_val)", DataTypes.SHORT.getName(), Literal.of(10));
+
+        assertEvaluate("pg_typeof(10::integer)", DataTypes.INTEGER.getName());
+        assertEvaluate("pg_typeof(a)", DataTypes.INTEGER.getName(), Literal.of(10));
+
+        assertEvaluate("pg_typeof(8765134432441)", "bigint");
+        assertEvaluate("pg_typeof(x)", DataTypes.LONG.getName(), Literal.of(8765134432441L));
+
+        assertEvaluate("pg_typeof(42.0::real)", DataTypes.FLOAT.getName());
+        assertEvaluate("pg_typeof(float_val)", DataTypes.FLOAT.getName(), Literal.of(42.0));
+
+        assertEvaluate("pg_typeof(42.0)", DataTypes.DOUBLE.getName());
+        assertEvaluate("pg_typeof(double_val)", DataTypes.DOUBLE.getName(), Literal.of(42.0));
+
+        assertEvaluate("pg_typeof('name')", DataTypes.STRING.getName());
+        assertEvaluate("pg_typeof(name)", DataTypes.STRING.getName(), Literal.of("name"));
+
+        assertEvaluate("pg_typeof('120.0.0.1'::ip)", DataTypes.IP.getName());
+        assertEvaluate("pg_typeof(ip)", DataTypes.IP.getName(), Literal.of("127.0.0.1"));
+
+        assertEvaluate("pg_typeof('1978-02-28T10:00:00+01'::timestamp with time zone)",
+                       DataTypes.TIMESTAMPZ.getName());
+        assertEvaluate("pg_typeof('1978-02-28T10:00:00+01'::timestamp without time zone)",
+                       DataTypes.TIMESTAMP.getName());
+        assertEvaluate("pg_typeof(timestamp)", "timestamp without time zone",
+                       Literal.of(DataTypes.TIMESTAMP, DataTypes.TIMESTAMP.value("1978-02-28T14:30+05:30")));
+        assertEvaluate("pg_typeof(timestamp_tz)", "timestamp with time zone",
+                       Literal.of(DataTypes.TIMESTAMPZ, DataTypes.TIMESTAMPZ.value("1978-02-28T14:30+05:30")));
+    }
+
+    @Test
+    public void test_primitive_types_with_qualified_function_name() {
+        assertEvaluate("pg_catalog.pg_typeof(null)", DataTypes.UNDEFINED.getName());
```

Comment:
I think this is a bit too much as we would have to validate only once that this function is reachable inside the `pg_catalog` namespace. The functionality of the function does not change.